src/tamm/proc_group.hpp

```diff
@@ -375,13 +375,36 @@ class ProcGroup {
 
   template<typename T>
   void allgather(const T* sbuf, T* rbuf) {
-    throw std::runtime_error("allgather unsupported");
+    EXPECTS(sbuf != nullptr);
+
+    auto             nranks = size().value();
+    upcxx::promise<> p(nranks);
+
+    for(int r = 0; r < nranks; ++r)
+      upcxx::broadcast(*sbuf, r, *team()).then([&rbuf, &p, r](T result) {
+        rbuf[r] = result;
+        p.fulfill_anonymous(1);
+      });
+
+    p.get_future().wait();
```

Comment:
This algorithm uses the value-based `broadcast()` overload. That is fine for scalars and other small types but could be very inefficient if `T` is a statically large type, because it repeatedly copies the value around on the stack.

It would be simpler to just replace this function body with a call to the vector variant:
```suggestion
allgather<T>(sbuf, 1, rbuf, 1);
```
which uses the vector broadcast overload, preserving efficiency in the case of large `T`.